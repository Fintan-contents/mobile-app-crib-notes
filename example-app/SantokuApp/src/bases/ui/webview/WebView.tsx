import {m} from 'bases/message/Message';
import React, {useCallback, useEffect, useState} from 'react';
import {ActivityIndicator, StyleSheet} from 'react-native';
import {WebView as RNWebView, WebViewProps} from 'react-native-webview';
import {
  WebViewErrorEvent,
  WebViewNavigationEvent,
  WebViewProgressEvent,
  WebViewScrollEvent,
  WebViewSource,
  WebViewSourceUri,
} from 'react-native-webview/lib/WebViewTypes';

import {Snackbar} from '../snackbar/Snackbar';

type Props = WebViewProps & {
  /**
   * Error message when React Native WebView original onError is raised.
   * If not specified, the default message is displayed.
   * If onError is specified by the parent, this message will not be displayed.
   */
  errorMessage?: string;
  /**
   * End-of-Scroll Event.
   * Occurs every time the web page scrolls to the bottom of the content.
   */
  onScrollEnd?: () => void;
  /**
   * End-of-scroll event that occurs only once.
   * Occurs only once when the web page scroll reaches the bottom of the content.
   */
  onScrollEndOnce?: () => void;
};

const isUriSource = (source?: WebViewSource): source is WebViewSourceUri => {
  return source !== undefined && 'uri' in source;
};

/**
 * WebViewでページを表示するコンポーネント。
 * react-native-webviewで提供されるイベントに、以下の2つのイベントを追加しています。
 * - onScrollEnd: スクロールがページの終端に移動する度に発生
 * - onScrollEndOnce: スクロールがページの終端に移動した最初の1回のみ発生
 *
 * 動作確認中に発生した以下の事象に対応しています。
 * - 発生OS
 *   - iOS
 * - 発生事象
 *   - URIが変わった直後にonLoadStart->onLoadEnd->onScroll->onScrollの順で発火し、1回目のonScrollで渡されるcontentOffsetが、URIが変わる前のページのcontentOffsetになっている。
 *     そのため、URIが変わる前のページで終端までスクロールしていた場合、1回目のonScrollで終端までスクロールされていると判定されてしまい、onScrollEnd/onScrollOnceイベントが発生してしまう。
 *     なお、2回目のonScrollで渡されるcontentOffsetは0になっている
 * - 対応方法
 *   - 1回目のonScrollでonScrollEnd/onScrollEndOnceイベントが発生しないように、URIが変わったかどうかを状態として持ち、
 *     URIが変わった直後の、オフセットが0ではないスクロールイベントでは、onScrollEnd/onScrollEndOnceは発生させない。
 *
 * - 発生OS
 *   - Android
 * - 発生事象
 *   - onLoadEndがページのロードが完了する前に呼び出される不具合がある。（https://github.com/react-native-webview/react-native-webview/issues/2345）
 * - 対応方法
 *   - onLoadProgressイベント発生時に、progressが1になったらページロードが完了したと判定する。
 *
 * なお、本コンポーネントには以下の制約事項があります。
 * - 対象OS
 *   - Android
 * - 制約事項
 *   - スクロールが存在しないページでは、onScrollEnd/onScrollEndOnceイベントは発生しません。
 *
 */
export const WebView = React.forwardRef<RNWebView, Props>(function WebView(props, ref) {
  const [loadEnd, setLoadEnd] = useState(false);
  const [scrollEndCalled, setScrollEndCalled] = useState(false);
  const {onScrollEnd, onScrollEndOnce, onLoadStart, onLoadProgress, onError, errorMessage, ...webViewProps} = props;

  const [isUriChanged, setIsUriChanged] = useState(false);
  const uri = isUriSource(props.source) ? props.source.uri : undefined;
  useEffect(() => {
    setIsUriChanged(true);
  }, [uri]);

  // uriが変わった時にWebViewがページを再ロードした場合の対応
  // なお、WebViewで表示しているページ内リンクでページ遷移した場合は、onLoadStartイベントは再発行されません
  const handleLoadStart = useCallback(
    (event: WebViewNavigationEvent) => {
      setLoadEnd(false);
      setScrollEndCalled(false);
      onLoadStart?.(event);
    },
    [onLoadStart],
  );

  const handleScroll = useCallback(
    (event: WebViewScrollEvent) => {
      if (isUriChanged && event.nativeEvent.contentOffset.y > 0) {
        // URLが変わった直後の、オフセットが0ではないスクロールイベントでは、onScrollEndは発生させない。
        setIsUriChanged(false);
      } else if ((onScrollEnd || onScrollEndOnce) && loadEnd) {
        // 小数点の誤差があるため、1px分は丸め誤差として扱う
        const scrollY = event.nativeEvent.contentOffset.y + event.nativeEvent.layoutMeasurement.height + 1;
        if (event.nativeEvent.contentSize.height <= scrollY) {
          if (!scrollEndCalled) {
            setScrollEndCalled(true);
            onScrollEndOnce?.();
          }
          onScrollEnd?.();
        }
      }

      props.onScroll?.(event);
    },
    [isUriChanged, loadEnd, scrollEndCalled, onScrollEnd, onScrollEndOnce, props],
  );

  const handleLoadProgress = useCallback(
    (event: WebViewProgressEvent) => {
      if (event.nativeEvent.progress === 1) {
        setLoadEnd(true);
      } else {
        setLoadEnd(false);
      }
      onLoadProgress?.(event);
    },
    [onLoadProgress],
  );

  const handleError = useCallback(
    (event: WebViewErrorEvent) => {
      if (onError) {
        onError(event);
      } else {
        Snackbar.showWithCloseButton(errorMessage ?? m('app.webview.onError'));
      }
    },
    [errorMessage, onError],
  );

  return (
    <RNWebView
      {...webViewProps}
      style={styles.container}
      onScroll={handleScroll}
      onLoadStart={handleLoadStart}
      onLoadProgress={handleLoadProgress}
      onError={handleError}
      ref={ref}
    />
  );
});

WebView.defaultProps = {
  // androidLayerType：WebViewの再レンダー時にクラッシュする可能性がある問題の回避策
  // https://github.com/react-native-webview/react-native-webview/issues/1915
  androidLayerType: 'software',
  startInLoadingState: true,
  // iOSでスクロールしたときに滑らかにスクロールするようにしておく。
  // https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#decelerationrate
  decelerationRate: 'normal',
  renderLoading: () => <ActivityIndicator style={styles.indicator} size="large" color="#0000ff" />,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  indicator: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
