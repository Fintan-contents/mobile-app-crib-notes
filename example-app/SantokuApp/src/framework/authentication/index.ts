export {AuthenticationService, ActiveAccountIdNotFoundError, PasswordNotFoundError} from './AuthenticationService';
export * from './SecureStorageAdapter';
export * from './UnauthorizedError';
