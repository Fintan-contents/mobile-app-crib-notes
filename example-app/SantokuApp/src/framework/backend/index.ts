export * from './refreshCsrfToken';
export * from './refreshSession';
