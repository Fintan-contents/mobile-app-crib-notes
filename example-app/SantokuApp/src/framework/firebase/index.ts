export {firebaseConfig, FirebaseConfig} from './FirebaseConfig';
