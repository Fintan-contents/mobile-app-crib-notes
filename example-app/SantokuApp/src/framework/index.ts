export {AuthenticationService, ActiveAccountIdNotFoundError, PasswordNotFoundError} from './authentication';
export {AppConfig} from './config';
export {firebaseConfig, FirebaseConfig} from './firebase';
export type {Logger, LoggerOptions} from './logging';
export {log, createLogger} from './logging';
export {m, loadMessages, BundledMessagesLoader} from './message';
export {LocalAuthentication} from './local-authentication';
export {launchedId, generatePassword, useIsMounted} from './utilities';
