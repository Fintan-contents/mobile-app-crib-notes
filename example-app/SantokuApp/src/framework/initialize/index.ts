export * from './withInitialData';
export * from './useAppInitializer';
