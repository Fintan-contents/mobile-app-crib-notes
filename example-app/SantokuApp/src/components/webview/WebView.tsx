import {useSnackbar} from 'components/overlay';
import {m} from 'framework';
import React, {useCallback, useState} from 'react';
import {ActivityIndicator, StyleSheet} from 'react-native';
import {WebView as RNWebView, WebViewProps} from 'react-native-webview';
import {WebViewErrorEvent, WebViewNavigationEvent, WebViewScrollEvent} from 'react-native-webview/lib/WebViewTypes';

type Props = WebViewProps & {
  /**
   * Error message when React Native WebView original onError is raised.
   * If not specified, the default message is displayed.
   * If onError is specified by the parent, this message will not be displayed.
   */
  errorMessage?: string;
  /**
   * End-of-Scroll Event.
   * Occurs every time the web page scrolls to the bottom of the content.
   */
  onScrollEnd?: () => void;
  /**
   * End-of-scroll event that occurs only once.
   * Occurs only once when the web page scroll reaches the bottom of the content.
   */
  onScrollEndOnce?: () => void;
};

export const WebView = React.forwardRef<RNWebView, Props>(function WebView(props, ref) {
  const [loadEnd, setLoadEnd] = useState(false);
  const [scrollEndCalled, setScrollEndCalled] = useState(false);
  const {onScrollEnd, onScrollEndOnce, ...webViewProps} = props;
  const snackbar = useSnackbar();

  const handleScroll = useCallback(
    (event: WebViewScrollEvent) => {
      if ((onScrollEnd || onScrollEndOnce) && loadEnd) {
        // 小数点の誤差があるため、1px分は丸め誤差として扱う
        const scrollY = event.nativeEvent.contentOffset.y + event.nativeEvent.layoutMeasurement.height + 1;
        if (event.nativeEvent.contentSize.height <= scrollY) {
          if (!scrollEndCalled) {
            setScrollEndCalled(true);
            onScrollEndOnce?.();
          }
          onScrollEnd?.();
        }
      }

      props.onScroll?.(event);
    },
    [loadEnd, scrollEndCalled, onScrollEnd, onScrollEndOnce, props],
  );

  const handleLoadEnd = useCallback(
    (event: WebViewNavigationEvent | WebViewErrorEvent) => {
      setLoadEnd(true);
      props.onLoadEnd?.(event);
    },
    [props],
  );

  const handleError = useCallback(
    (event: WebViewErrorEvent) => {
      if (props.onError) {
        props.onError(event);
      } else {
        snackbar.showWithCloseButton(props.errorMessage ?? m('app.webview.onError'));
      }
    },
    [props, snackbar],
  );

  return (
    <RNWebView
      {...webViewProps}
      style={styles.container}
      onScroll={handleScroll}
      onLoadEnd={handleLoadEnd}
      onError={handleError}
      ref={ref}
    />
  );
});

WebView.defaultProps = {
  // androidLayerType：WebViewの再レンダー時にクラッシュする可能性がある問題の回避策
  // https://github.com/react-native-webview/react-native-webview/issues/1915
  androidLayerType: 'software',
  startInLoadingState: true,
  // iOSでスクロールしたときに滑らかにスクロールするようにしておく。
  // https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#decelerationrate
  decelerationRate: 'normal',
  renderLoading: () => <ActivityIndicator style={styles.indicator} size="large" color="#0000ff" />,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  indicator: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
