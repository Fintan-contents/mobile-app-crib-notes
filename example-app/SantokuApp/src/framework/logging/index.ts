import {firebaseConfig} from '../firebase';
import {FirebaseCrashlyticsTransport} from './FirebaseCrashlyticsTransport';
import {createLogger, Logger, LoggerOptions} from './Logger';

const log =
  __DEV__ || firebaseConfig.isDummy
    ? createLogger({level: 'trace'})
    : createLogger({level: 'error', transports: [new FirebaseCrashlyticsTransport()]});

export type {Logger, LoggerOptions};
export {log, createLogger};
