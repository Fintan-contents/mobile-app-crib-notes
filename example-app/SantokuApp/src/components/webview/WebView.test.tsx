import {fireEvent, render} from '@testing-library/react-native';
import {BundledMessagesLoader, loadMessages} from 'framework';
import React from 'react';

import {WebView} from './WebView';

const mockedShowWithCloseButton = jest.fn();

jest.mock('components/overlay', () => ({
  useSnackbar: () => ({
    showWithCloseButton: mockedShowWithCloseButton,
  }),
}));

describe('WebView', () => {
  it('WebViewが正常にrenderできることを確認', () => {
    const renderResult = render(<WebView testID="webview" source={{uri: 'https://localhost/'}} />);
    expect(renderResult.queryByTestId('webview')).not.toBeNull();
    expect(renderResult).toMatchSnapshot();
  });

  it('WebViewのスクロールイベントが適切なタイミングで発行されることを確認', () => {
    const handleOnScroll = jest.fn();
    const handleOnScrollEnd = jest.fn();
    const handleOnScrollEndOnce = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    const webview = render(
      <WebView
        source={{uri: 'https://localhost/'}}
        onScroll={handleOnScroll}
        onScrollEnd={handleOnScrollEnd}
        onScrollEndOnce={handleOnScrollEndOnce}
        testID="webview"
      />,
    ).getByTestId('webview');

    expect(webview).not.toBeNull();

    // onLoadEndイベント発生前はonScrollEndイベントを発行しない
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(1);
    expect(handleOnScrollEnd).not.toHaveBeenCalled();
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    fireEvent(webview, 'onLoadEnd');

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(2);
    expect(handleOnScrollEnd).not.toHaveBeenCalled();
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // スクロール終端と判定（１回目）
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(3);
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(1);
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);

    // スクロール終端と判定（２回目のため、onScrollEndOnceイベントは発生しない）
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(4);
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(2);
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);
  });

  it('WebViewにonScrollEndのみ指定して正常にイベントが発行されることを確認', () => {
    const handleOnScrollEnd = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    const webview = render(
      <WebView source={{uri: 'https://localhost/'}} onScrollEnd={handleOnScrollEnd} testID="webview" />,
    ).getByTestId('webview');

    fireEvent(webview, 'onLoadEnd');

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEnd).not.toHaveBeenCalled();

    // スクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(1);
  });

  it('WebViewにonScrollEndOnceのみ指定して正常にイベントが発行されることを確認', () => {
    const handleOnScrollEndOnce = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    const webview = render(
      <WebView source={{uri: 'https://localhost/'}} onScrollEndOnce={handleOnScrollEndOnce} testID="webview" />,
    ).getByTestId('webview');

    fireEvent(webview, 'onLoadEnd');

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // スクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);
  });

  it('WebViewのonErrorイベントで、親からonErrorを指定してない場合はSnackbar表示関数をコールすることを確認', async () => {
    await loadMessages(new BundledMessagesLoader());

    const webview = render(<WebView source={{uri: 'https://localhost/'}} testID="webview" />).getByTestId('webview');

    fireEvent(webview, 'onError');
    expect(mockedShowWithCloseButton).toHaveBeenCalledTimes(1);
  });

  it('WebViewのonErrorイベントで、親からonErrorを指定した場合は親へイベント通知することを確認', () => {
    const handleOnError = jest.fn();

    const webview = render(
      <WebView source={{uri: 'https://localhost/'}} onError={handleOnError} testID="webview" />,
    ).getByTestId('webview');

    fireEvent(webview, 'onError');
    expect(handleOnError).toHaveBeenCalledTimes(1);
  });
});
