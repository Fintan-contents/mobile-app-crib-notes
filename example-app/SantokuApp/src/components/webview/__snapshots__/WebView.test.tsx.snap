// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebView WebViewが正常にrenderできることを確認 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    androidLayerType="software"
    cacheEnabled={true}
    decelerationRate={0.998}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={false}
    onContentProcessDidTerminate={[Function]}
    onError={[Function]}
    onHttpError={[Function]}
    onLoadEnd={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onScroll={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      Object {
        "uri": "https://localhost/",
      }
    }
    startInLoadingState={true}
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="webview"
    useSharedProcessPool={true}
  />
  <ActivityIndicator
    animating={true}
    color="#0000ff"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  />
</View>
`;
